prow/tide/tide.go

```diff
@@ -357,15 +365,25 @@ func (sc *statusController) setStatuses(all, pool []PullRequest) {
 
 	process := func(pr *PullRequest) {
 		processed.Insert(prKey(pr))
-
 		log := sc.logger.WithFields(pr.logFields())
 		contexts, err := headContexts(log, sc.ghc, pr)
 		if err != nil {
 			log.WithError(err).Error("Getting head commit status contexts, skipping...")
 			return
 		}
-
-		wantState, wantDesc := expectedStatus(queriesByRepo, pr, poolM)
+		cr := newContextRegister(statusContext)
+		if sc.ca.Config().Tide.SkipOptionalContexts {
+			bp, err := sc.ca.Config().GetBranchProtection(
+				string(pr.Repository.Owner.Login),
+				string(pr.Repository.Name),
+				string(pr.BaseRef.Name))
+			if err != nil {
+				log.WithError(err).Error("Getting branch protection")
+				return
+			}
+			cr.registerRequiredContexts(findRequiredContexts(bp)...)
```

Comment:
It doesn't seem right that all of this is conditional on `sc.ca.Config().Tide.SkipOptionalContexts`.
If `SkipOptionalContexts` is false we should still require the required contexts.
Additionally, it doesn't look like we are ever actually adding the optional contexts with `registerOptionalContexts`.

Reply:
Should we get ride of the SkipOptionalContexts all together ?

The way it is implemented now, is that once required checks are registered all other checks are required optional, so you don't need to register the optional ones.

Reply:
It doesn't provide all of the behavior cases that we need to support. I think what we really want is a `SkipUnknownContexts` instead. 
As this PR is currently, if we require that all statuses are green to merge (including ephemeral ones that **might** be on a PR) by setting `SkipOptionalContexts=false` we wouldn't be able to require/expect specific statuses. This is problematic because this is the default value for a Tide config and will likely be the common case for smaller deployments since requiring all statuses to be green to merge is simple and intuitive.

We should change the contextRegister to have a set of required contexts, a set of known optional contexts (`tide` and **jobs marked optional**) and a bool indicating if unknown statuses should be required (`SkipUnknownContexts`).

When we are checking if a context can be ignored we just check that it is either 
1. in the optional set
or
1. it is not in the required set and `SkipUnknownContexts` is true.

When we are checking for missing required contexts we just ensure that all the contexts in the required set are present.

WDYT?

Reply:
I think there is a misunderstanding, if SkipOptionalContexts=false it is exactly as it is today. the only optional check will be tide and all other will be considered required.

Reply:
https://sourcegraph.com/github.com/sebastienvas/k8s-test-infra@7b4ef04aa5779be59c91d555fb697f8339d81c7f/-/blob/prow/tide/contextregister.go#L70